lib/spack/spack/spec.py

```diff
@@ -3439,6 +3429,14 @@ def normalize(self, force=False, tests=False, user_spec_deps=None):
 
         any_change = self._normalize_helper(visited, all_spec_deps, provider_index, tests)
 
+        # remove any leftover dependents outside the spec from, e.g., pruning externals
+        valid = {id(spec) for spec in all_spec_deps.values()} | {id(self)}
+        for spec in all_spec_deps.values():
+            remove = [dep for dep in spec.dependents() if id(dep) not in valid]
+            for dep in remove:
+                del spec._dependents.edges[dep.name]
+                del dep._dependencies.edges[spec.name]
```

Comment:
OT (comment, no action needed): I noticed slight variations of this pattern in many places in the code. I introduced in https://github.com/spack/spack/pull/40636 a class that helps keeping consistency between objects in memory and specs. See the PR for three similar cases that have been unified in the code where we want to ensure that a set of specs doesn't refere to external specs, and have the same `id` in memory.